Source/Urho3D/CSharp/Managed/Scene/IndexBuffer.cs

```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Urho3DNet
+{
+    public partial class IndexBuffer
+    { 
+        public bool SetData(short[] indexData)
+        {
+            //// Initialize unmanaged memory to hold the array.
+            int indexDatasize = Marshal.SizeOf(indexData[0]) * indexData.Length;
+            IntPtr indexDatapnt = Marshal.AllocHGlobal(indexDatasize);
+            return SetData(indexDatapnt);
```

Comment:
Very useful addition. There are multiple issues though.

1. `indexData` can be null.
2. `indexDatapnt` name is odd and does not follow naming conventions. `indexDatasize` also does not follow name conventions.
3. `indexDatapnt` is never freed.
4. `indexData` is not copied into `indexDatapnt`.

For best performance you should pin `indexData` using `fixed` statement and pass pinned pointer to `SetData()`. You will avoid memory allocation and copying. `IndexBuffer` copies data on it's own.

Reply:
ooh!... looks like i missed this lesson. it is good to refresh myself. now going through `fixed` I realized `it is not possible to make an unmanaged pointer of a managed object or of a struct that contains managed fields.`
so far i resolved like this
```
 public bool SetData(short[] indexData)
        {  
           if(  indexData==null || indexData.Length==0) return false;
            //// Initialize unmanaged memory to hold the array.
            int indexDataSize = Marshal.SizeOf(indexData[0]) * indexData.Length;
            IntPtr indexDataPtr = Marshal.AllocHGlobal(indexDataSize);
            Marshal.StructureToPtr(indexData, indexDataPtr, true);

            var result = SetData(indexDataPtr);
            Marshal.FreeHGlobal(indexDataPtr);

            return result;
        }
```
not sure shall i `try-catch` it with a finally `Marshal.FreeHGlobal(indexDataPtr);` to avoid any possibility of memory leak?

Reply:
Not sure what you mean here. `indexData` is not a struct so you can pin it just fine:
```cs
fixed (short* indexDataPtr = indexData)
{
    // ...
}
```

Reply:
(I could be wrong)  I recall since we added a managed fields to the short[], it is not possible to make an unmanaged pointer
[see this](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.structuretoptr?redirectedfrom=MSDN&view=netframework-4.7.1#System_Runtime_InteropServices_Marshal_StructureToPtr_System_Object_System_IntPtr_System_Boolean_)
however, I sorted it like this
```
 unsafe public bool SetData(short[] indexData)
        {
            fixed (short* indexDataPtr = indexData)
            {
                return SetData((IntPtr)indexDataPtr );
            }
        }
```

Reply:
What managed fields to `short[]`? `short[]` is an array primitives. By definition you can not add any fields to an array of plain numbers.

Reply:
aha.. thanks... then it should be fine. I think i mixed stuff. I will read more about it.